File: solutions/swift/pig-latin/1/Sources/PigLatin/PigLatin.swift
class PigLatin {

    static func translate(_ word: String) -> String {
        let words = word.split(separator: " ").map { String($0) }
        if words.count > 1 {
            return words.map { translateWord($0) }.joined(separator: " ")
        }
        return translateWord(word)
    }

    private static func leadingConsonantCluster(_ word: String) -> String {
        let vowels: [Character] = ["a", "e", "i", "o", "u"]
        var cluster = ""
        for (i, character) in word.enumerated() {
            if vowels.contains(character) { break }
            if character == "y" && i > 0 { break }
            cluster.append(character)
            if character == "q" && word.dropFirst(i + 1).first == "u" {
                cluster.append("u")
                break
            }
        }
        return cluster
    }

    private static func translateWord(_ word: String) -> String {
        let vowels: [Character] = ["a", "e", "i", "o", "u"]
        let startsWithVowelPair = ["xr", "yt"]

        if let first = word.first, vowels.contains(first) {
            return word + "ay"
        }

        for pair in startsWithVowelPair {
            if word.hasPrefix(pair) {
                return word + "ay"
            }
        }

        let leadingCluster = leadingConsonantCluster(word)
        let core = String(word.dropFirst(leadingCluster.count))
        return core + leadingCluster + "ay"
    }
}

File: solutions/ruby/pig-latin/1/pig_latin.rb
class PigLatin

  VOWELS = %w[a e i o u]
  UNUSUAL_VOWEL_PAIRS = %w[xr yt]
  AY = 'ay'
  private_constant :VOWELS, :UNUSUAL_VOWEL_PAIRS, :AY

  def self.leading_consonant_cluster(word)
    cluster = ''
    word.chars.each_with_index do |character, i|
      break if VOWELS.include?(character)
      break if character == 'y' && i > 0
      cluster += character
      if character == 'q' && word[i + 1] == VOWELS.last
        cluster += VOWELS.last
        break
      end
    end

    cluster
  end

  def self.translate_word(word)
    return word + AY if VOWELS.include?(word[0])
    return word + AY if UNUSUAL_VOWEL_PAIRS.any? { |pair| word.start_with?(pair) }

    cluster = leading_consonant_cluster(word)
    core = word[cluster.length..]
    core + cluster + AY
  end
  private_class_method :leading_consonant_cluster, :translate_word

  public

  def self.translate(words)
    words.split(' ').map { |word| translate_word(word) }.join(' ')
  end

end
